extensions/interactions/CodeRepl/CodeRepl.js

```diff
@@ -146,3 +146,140 @@ oppia.directive('oppiaShortResponseCodeRepl', [
     };
   }
 ]);
+
+oppia.factory('codeReplNormalizationService', [function() {
+  var removeLeadingWhitespace = function(str) {
+    return str.replace(/^\s+/g, '');
+  };
+  var removeTrailingWhitespace = function(str) {
+    return str.replace(/\s+$/g, '');
+  };
+  return {
+    getNormalizedCode: function(codeString) {
+      /*
+       * Normalizes a code string (which is assumed not to contain tab
+       * characters). In particular:
+       *
+       * - Strips out lines that start with '#' (comments), possibly preceded by
+       *     whitespace.
+       * - Trims trailing whitespace on each line.
+       * - Removes blank newlines.
+       * - Make the indentation level four spaces.
+       */
+      // TODO(sll): Augment this function to strip out comments that occur at the
+      // end of a line. However, be careful with lines where '#' is contained in
+      // quotes or the character is escaped.
+      var FOUR_SPACES = '    ';
+      // Maps the number of spaces at the beginning of a line to an int specifying
+      // the desired indentation level.
+      var numSpacesToDesiredIndentLevel = {
+          0: 0,
+      };
+
+      var codeLines = removeTrailingWhitespace(codeString).split('\n');
+      var normalizedCodeLines = [];
+      codeLines.forEach(function(line) {
+        if (removeLeadingWhitespace(line).indexOf('#') === 0) {
+          return;
+        }
+        line = removeTrailingWhitespace(line);
+        if (!line) {
+          return;
+        }
+
+        var numSpaces = line.length - removeLeadingWhitespace(line).length;
+
+        var existingNumSpaces = Object.keys(numSpacesToDesiredIndentLevel);
+        var maxNumSpaces = Math.max.apply(null, existingNumSpaces);
+        if (numSpaces > maxNumSpaces) {
+          // Add a new indentation level
+          numSpacesToDesiredIndentLevel[numSpaces] = existingNumSpaces.length;
+        }
+
+        // This is set when the indentation level of the current line does not
+        // start a new scope, and also does not match any previous indentation
+        // level. This case is actually invalid, but for now, we take the
+        // largest indentation level that is less than this one.
+        // TODO(sll): Bad indentation should result in an error nearer the
+        // source.
+        var isShortfallLine =
+          !numSpacesToDesiredIndentLevel.hasOwnProperty(numSpaces) &&
+          numSpaces < maxNumSpaces;
+
+        // Clear all existing indentation levels to the right of this one.
+        for (var indentLength in numSpacesToDesiredIndentLevel) {
+          if (Number(indentLength) > numSpaces) {
+            delete numSpacesToDesiredIndentLevel[indentLength];
+          }
+        }
+
+        if (isShortfallLine) {
+          var existingIndentLevels = existingNumSpaces.map(function(indentLength) {
+            return numSpacesToDesiredIndentLevel[indentLength];
+          });
+          numSpaces = Math.max.apply(null, existingIndentLevels);
```

Comment:
I'm confused about this. Shouldn't this be the max of the keys, not of the values?